src/client/clientmap.cpp

```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+				occlusion_culling_enabled = false;
+		}
+
+		v3s16 camera_block = getContainerPos(cam_pos_nodes, MAP_BLOCKSIZE);
+		m_drawlist = std::map<v3s16, MapBlock*, MapBlockComparer>(MapBlockComparer(camera_block));
+
+		auto is_frustum_culled = m_client->getCamera()->getFrustumCuller();
+
+		// Uncomment to debug occluded blocks in the wireframe mode
+		// TODO: Include this as a flag for an extended debugging setting
+		//if (occlusion_culling_enabled && m_control.show_wireframe)
+		//    occlusion_culling_enabled = porting::getTimeS() & 1;
+
+		for (const auto &sector_it : m_sectors) {
+			MapSector *sector = sector_it.second;
+			v2s16 sp = sector->getPos();
+
+			blocks_loaded += sector->size();
+			if (!m_control.range_all) {
+				if (sp.X < p_blocks_min.X || sp.X > p_blocks_max.X ||
+						sp.Y < p_blocks_min.Z || sp.Y > p_blocks_max.Z)
+					continue;
+			}
+
+			MapBlockVect sectorblocks;
+			sector->getBlocks(sectorblocks);
+
+			/*
+				Loop through blocks in sector
+			*/
+
+			u32 sector_blocks_drawn = 0;
+
+			for (MapBlock *block : sectorblocks) {
+				/*
+					Compare block position to camera position, skip
+					if not seen on display
+				*/
+
+				if (!block->mesh) {
+					// Ignore if mesh doesn't exist
+					continue;
+				}
+
+				v3s16 block_coord = block->getPos();
+				v3f mesh_sphere_center = intToFloat(block->getPosRelative(), BS)
+						+ block->mesh->getBoundingSphereCenter();
+				f32 mesh_sphere_radius = block->mesh->getBoundingRadius();
+				// First, perform a simple distance check.
+				if (!m_control.range_all &&
+					mesh_sphere_center.getDistanceFrom(intToFloat(cam_pos_nodes, BS)) >
+						m_control.wanted_range * BS + mesh_sphere_radius)
+					continue; // Out of range, skip.
+
+				// Keep the block alive as long as it is in range.
+				block->resetUsageTimer();
+				blocks_in_range_with_mesh++;
+
+				// Frustum culling
+				// Only do coarse culling here, to account for fast camera movement.
+				// This is needed because this function is not called every frame.
+				constexpr float frustum_cull_extra_radius = 300.0f;
+				if (is_frustum_culled(mesh_sphere_center,
+						mesh_sphere_radius + frustum_cull_extra_radius))
+					continue;
+
+				// Occlusion culling
+				if (occlusion_culling_enabled && isBlockOccluded(block, cam_pos_nodes)) {
+					blocks_occlusion_culled++;
+					continue;
+				}
+
+				// Add to set
+				block->refGrab();
+				m_drawlist[block_coord] = block;
+
+				sector_blocks_drawn++;
+			} // foreach sectorblocks
+
+			if (sector_blocks_drawn != 0)
+				m_last_drawn_sectors.insert(sp);
+		}
+
+		g_profiler->avg("MapBlock meshes in range [#]", blocks_in_range_with_mesh);
+		g_profiler->avg("MapBlocks occlusion culled [#]", blocks_occlusion_culled);
+		g_profiler->avg("MapBlocks drawn [#]", m_drawlist.size());
+		g_profiler->avg("MapBlocks loaded [#]", blocks_loaded);
 	}
-	m_drawlist.clear();
+}
+
+void ClientMap::touchMapBlocks()
```

Comment:
This was the side effect of the previous algorithm to keep loaded blocks in range warm to avoid reloads from server. I only kept existing functionality.